presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/PushRemoteExchangeThroughGroupId.java

```diff
@@ -0,0 +1,123 @@
+ */
+package com.facebook.presto.sql.planner.iterative.rule;
+
+import com.facebook.presto.Session;
+import com.facebook.presto.SystemSessionProperties;
+import com.facebook.presto.matching.Capture;
+import com.facebook.presto.matching.Captures;
+import com.facebook.presto.matching.Pattern;
+import com.facebook.presto.spi.relation.VariableReferenceExpression;
+import com.facebook.presto.sql.planner.Partitioning;
+import com.facebook.presto.sql.planner.PartitioningScheme;
+import com.facebook.presto.sql.planner.iterative.Rule;
+import com.facebook.presto.sql.planner.plan.ExchangeNode;
+import com.facebook.presto.sql.planner.plan.GroupIdNode;
+import com.google.common.collect.ImmutableList;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.Set;
+import java.util.stream.Collectors;
+
+import static com.facebook.presto.matching.Capture.newCapture;
+import static com.facebook.presto.sql.planner.plan.ExchangeNode.Type.REPLICATE;
+import static com.facebook.presto.sql.planner.plan.Patterns.exchange;
+import static com.facebook.presto.sql.planner.plan.Patterns.groupId;
+import static com.facebook.presto.sql.planner.plan.Patterns.source;
+import static com.google.common.collect.ImmutableList.toImmutableList;
+import static com.google.common.collect.Iterables.getOnlyElement;
+
+public final class PushRemoteExchangeThroughGroupId
+        implements Rule<ExchangeNode>
+{
+    private static final Capture<GroupIdNode> GROUP_ID = newCapture();
+    private static final Pattern<ExchangeNode> PATTERN = exchange()
+            .matching(exchange -> exchange.getScope().isRemote())
+            .matching(exchange -> exchange.getType() != REPLICATE)
+            .with(source().matching(
+                    groupId().capturedAs(GROUP_ID)));
+    @Override
+    public Pattern<ExchangeNode> getPattern()
+    {
+        return PATTERN;
+    }
+
+    @Override
+    public boolean isEnabled(Session session)
+    {
+        return SystemSessionProperties.shouldPushRemoteExchangeThroughGroupId(session);
+    }
+
+    @Override
+    public Result apply(ExchangeNode node, Captures captures, Context context)
+    {
+        GroupIdNode groupIdNode = captures.get(GROUP_ID);
+        Set<VariableReferenceExpression> commonGroupingColumns = groupIdNode.getCommonGroupingColumns();
+        if (commonGroupingColumns.isEmpty()) {
+            // If there is no common grouping column we can not push exchange operator below groupId.
+            return Result.empty();
+        }
+
+        PartitioningScheme partitioningScheme = node.getPartitioningScheme();
+        List<VariableReferenceExpression> inputs = getOnlyElement(node.getInputs());
+        inputs = removeVariable(inputs, groupIdNode.getGroupIdVariable());
+        inputs = replaceAlias(inputs, groupIdNode.getGroupingColumns());
+
+        List<VariableReferenceExpression> outputLayout = partitioningScheme.getOutputLayout();
```

Comment:
Do you think it makes sense to move line 75 to above line 80, so that 'partitioningScheme' is close to the place where it is firstly used?

Reply:
Done